File: scripting/engine/Playground/ecs/components/BoxColliderComponent.swift
@_alignment(8)
@frozen
@component
public struct BoxColliderComponent {
    public var dimensions: Vector3 = .zero
    public var offset: Vector3 = .zero
    public var rotation: Quaternion = .identity
    public var material: UnsafeMutableRawPointer
    internal var handle: UInt64 = UInt64.max
    internal var bodyHandle: UInt64 = UInt64.max
    public var isTrigger: Bool = false

    public init(isTrigger: Bool, dimensions: Vector3, offset: Vector3, rotation: Quaternion, material: UnsafeMutableRawPointer) {
        self.isTrigger = isTrigger
        self.dimensions = dimensions
        self.offset = offset
        self.rotation = rotation
        self.material = material
    }
}

File: engine/managed/PlaygroundAssembly/PlaygroundAssembly/ECS/Physics/BoxColliderComponent.cs
using System.Numerics;
using System.Runtime.InteropServices;
using Playground.Core.Ecs;
using PlaygroundAssembly.ECS.Core;
using Vector3 = PlaygroundAssembly.Math.Vector3;

namespace PlaygroundAssembly.ECS.Physics;

[EcsComponent]
[StructLayout(LayoutKind.Explicit, Pack = 16, Size = 80)]
public struct BoxColliderComponent(
    bool isTrigger,
    Vector3 dimensions,
    Vector3 offset,
    Quaternion rotation,
    IntPtr material)
{
    [FieldOffset(0)]
    public Vector3 Dimensions = dimensions;
    [FieldOffset(12)]
    public Vector3 Offset = offset;
    [FieldOffset(24)]
    public Quaternion Rotation = rotation;
    [FieldOffset(40)]
    public IntPtr Material = material;
    // Internal runtime handles
    [FieldOffset(48)]
    internal ulong Handle = ulong.MaxValue;
    [FieldOffset(56)]
    internal ulong BodyHandle = ulong.MaxValue;
    [FieldOffset(64)]
    public bool IsTrigger = isTrigger;
}
